src/RazorSdk/Tool/DiscoverCommand.cs

```diff
@@ -89,14 +89,48 @@ protected override bool ValidateArguments()
                 }
             }
 
+            PatchExtensions(ExtensionNames, ExtensionFilePaths);
+
             return true;
         }
 
+        /// <summary>
+        /// Replaces the assembly for MVC extension v1 or v2 with the one shipped alongside SDK (as opposed to the one from NuGet).
+        /// </summary>
+        /// <remarks>
+        /// Needed so the Razor compiler can change its APIs without breaking legacy MVC scenarios.
+        /// </remarks>
+        internal static void PatchExtensions(CommandOption extensionNames, CommandOption extensionFilePaths)
+        {
+            string currentDirectory = null;
+
+            for (int i = 0; i < extensionNames.Values.Count; i++)
+            {
+                var replacementFileName = extensionNames.Values[i] switch
+                {
+                    "MVC-1.0" or "MVC-1.1" => "Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X.dll",
+                    "MVC-2.0" or "MVC-2.1" => "Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X.dll",
+                    _ => null,
+                };
+
+                if (replacementFileName != null && HasExpectedFileName(extensionFilePaths.Values[i]))
```

Comment:
Should we write a warning if the `replacmentFileName != null` but it does not have the expected name? Can't think of a scenario where that would make sense.